File: 22/part2.rb
require 'set'

a, b = File.read('./input.txt').split("\n\n").map do |chunk|
  chunk.split("\n").drop(1).map(&:to_i)
end

def play(x, y)
  winner = nil
  memory = Set.new

  until x.empty? || y.empty?
    return :x if memory.include?([x,y])

    memory << [x.dup, y.dup]
    xx, yy = x.shift, y.shift

    winner =
      if xx <= x.length && yy <= y.length
        play(x.take(xx), y.take(yy))
      else
        xx > yy ? :x : :y
      end

    winner == :x ? x << xx << yy : y << yy << xx
  end

  winner
end

play(a, b)

puts([a, b].detect(&:any?).reverse.zip(1...).reduce(0) {|m,(c,i)| m + c*i})

File: 22/part2.swift
import Foundation

let file = String(data: FileManager.default.contents(atPath: "./input.txt")!, encoding: .utf8)!
let deck = file.components(separatedBy: "\n\n").map { chunk in
  chunk.split(separator: "\n").dropFirst(1).map { Int($0)! }
}

var a = deck[0],
    b = deck[1]


@discardableResult func play(_ x: inout [Int], _ y: inout [Int]) -> Int {
  var winner = 0
  var memory = Set<[[Int]]>()

  while !x.isEmpty && !y.isEmpty {
    if memory.contains([x,y]) {
      return -1
    }

    memory.insert([Array(x), Array(y)])

    let xx = x.removeFirst(),
        yy = y.removeFirst()

    if xx <= x.count && yy <= y.count {
      var subX = Array(x.prefix(xx)),
          subY = Array(y.prefix(yy))
      winner = play(&subX, &subY)
    } else {
      winner = xx > yy ? -1 : 1
    }

    if winner == -1 {
      x.append(xx)
      x.append(yy)
    } else {
      y.append(yy)
      y.append(xx)
    }
  }
  return winner
}

play(&a, &b)

let winner = [a,b].first(where: { !$0.isEmpty})!
print(zip(winner.reversed(), 1...).reduce(0, {m, a in
  let (b, c) = a
  return m + b * c
}))
